---
title: "Run BayEnv2"
output:
  html_notebook:
      toc: true 
      toc_float: true
      number_sections: True
      theme: lumen
      highlight: tango
      code_folding: hide
      df_print: paged
---
```{r eval=FALSE, message=FALSE, warning=FALSE}
library(gridExtra)
#library(scales)
source("../Rscripts/BaseScripts.R")
library(data.table)
library(DataCombine)
```

# Prepare SNPs for BayEnv
## Prune the VCF file with the strict standard 


```{bash eval=FALSE, message=FALSE, warning=FALSE}
#at farm (prune_vcf_froBayEnv.sh)
module load plink
#add variant ID names for pruning 
plink --vcf /home/ktist/ph/data/new_vcf/MD7000/PH_DP600_7000_minQ20_minMQ30_NS0.5_maf05.vcf.gz --set-missing-var-ids @:#[ph] --make-bed --out /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/MD7000_maf0.05 

plink --bfile /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/MD7000_maf0.05  --recode --tab --out /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/MD7000_maf0.05 

plink --file  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/MD7000_maf0.05  --indep-pairwise 1000kb 1 0.15 --out /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/prune_strict_1000k_0.15

#Different parameters (window size and R2 number) did not change the number of pruned loci drastically:
# indep-pairwise 50 5 0.5 (original parameters) = 270000 loci
# stricter parameters 1000k 1 0.15   = 205726

# Filter the vcf file (prune_convert_BAYENV.sh)
# First, reformat the prune.in  file as a 'CHR' 'POS' tab-delimited txt 
sed -e s/\\[ph\\]//g prune_strict_1000k_0.15.prune.in | awk  '{gsub(":","\t",$0); print;}' > prune_strict_1000k_0.15.in.txt

# Remove the pruned loci
bcftools view -R /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/prune_strict_1000k_0.15.in.txt /home/ktist/ph/data/new_vcf/MD7000/PH_DP600_7000_minQ20_minMQ30_NS0.5_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf

bgzip /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf


# Subset the each file (subset_prunedVCF2.sh)
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/BC17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_BC17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_BC17_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/CA17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_CA17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_CA17_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/PWS07.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS07_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS07_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/PWS17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS17_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/PWS91.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS91_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS91_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/PWS96.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS96_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS96_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/SS06.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS06_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS06_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/SS17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS17_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/SS96.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS96_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_SS96_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/TB06.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB06_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB06_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/TB17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB17_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/TB91.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB91_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB91_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/TB96.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB96_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_TB96_maf05.vcf.gz 
bcftools view -Oz -S /home/ktist/ph/data/new_vcf/population/WA17.txt --threads 24  /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_strict_PH_all_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_WA17_maf05.vcf.gz
bcftools index /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_WA17_maf05.vcf.gz 

#Count the alleles (extract_AC.sh)
bcftools query -f '%CHROM  %POS  %INFO/AC  %INFO/AN\n' /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS17_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_AC_PWS17.txt 
bcftools query -f '%CHROM  %POS  %INFO/AC  %INFO/AN\n' /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS91_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_AC_PWS91.txt 
bcftools query -f '%CHROM  %POS  %INFO/AC  %INFO/AN\n' /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/Pruned_PWS96_maf05.vcf.gz > /home/ktist/ph/data/new_vcf/MD7000/plinkfiles/pruned_AC_PWS96.txt 

```



## Create Allele Count file 
* Run extract_AC.sh to get the allele counts for each population at each loci
* Need to be filtered to unlinked loci

```{r eval=FALSE, message=FALSE, warning=FALSE}
pops.info<-read.csv("../Data/Sample_metadata_892pops.csv")
pops<-unique(pops.info$Population.Year)



#Pruned position info
prun<-read.table("../Data/new_vcf/AC/Pruned_loci.txt")
prun$V1<-paste0("chr",prun$V1)
colnames(prun)<-c("chr","pos")

snpfile<-data.frame(n=1:(2*nrow(prun)))
for (i in 1:length(pops)){
    df<-read.table(paste0("../Data/new_vcf/AC/AC_",pops[i],".txt"))
    colnames(df)<-c("chr","pos","AC","AN")
     # filter to only pruned positions
    df<-merge(df,prun, by=c("chr","pos"))
    #majorAC data frame
    maj<-transform(df, AC=AN-AC)
    newdf<-rbind(maj[,1:3], df[,1:3])
    #reorder the rows
    n<-nrow(df)
    newdf2<-newdf[kronecker(1:n, c(0, n), "+"), ]
    newdf2<-newdf2[order(newdf2$chr,newdf2$pos),]
    colnames(newdf2)[3]<-pops[i]
    #combine the data
    if (i==1) snpfile<-cbind(snpfile,newdf2)
    if (i!=1) snpfile<-cbind(snpfile, newdf2[,3])
}

colnames(snpfile)[5:17]<-pops[2:14]
#save the snpfi\e
snps<-snpfile[,-1]
write.csv(snps, "../Output/BayEnv/PH_allPops_snpsfile.csv")
write.table(snps[,3:16], "../Output/BayEnv/PH_allPops_snpsfile.txt", sep="\t", row.names = F,col.names = F, quote=F)

#Create year 2017 only file
snps<-read.csv("../Output/BayEnv/PH_allPops_snpsfile.csv", row.names = 1)
snps17<-snps[,grep("17",colnames(snps))]

#remove the loci with minor allele count=0  
minorzero<-which(rowSums(snps17)==0)
#check if minorzero are even or odd numbers
even<-minorzero[sapply(minorzero, function(x) x%%2==0)] 

# they are all odd numbers, so remove the odd line +1 from the snp file
remove<-c(minorzero, minorzero+1)
snps17_2<-snps17[-remove,]
nrow(snps17_2[which(rowSums(snps17_2)==0),])

write.table(snps17_2, "../Output/BayEnv/Y2017_Pops_snpsfile.txt", sep="\t", row.names = F,col.names = F, quote=F)

```

# Env data  

```{r eval=FALSE, message=FALSE, warning=FALSE}
# data contains missing numbers
df<-read.csv("../Data/gak1-mooring_PWS_temp.csv", skip=1)
depth<-unique(df$m)

re<-data.frame(depth=depth)
for (i in 1:length(depth)){
    sub<-df[df$m==depth[i],]
    re$nan[i]<-nrow(sub[is.nan(sub$degree_Celsius),])
    
}
# the shollowest temperature with no-missing data is 35m

#library(lubridate)
temp<-df[df$m==-35,]
temp$date<-as.Date(temp$UTC)
#temp$time<-parse_date_time(temp$time.UTC, "YmdHMS", tz="UTC")

temp<-as.data.table(temp)


#meanTemp<-setDT(temp)[, .(mean_temp = mean(degree_Celsius, na.rm=T)), by = .(year = year(date), month = months(date))]

#meanTemp$date<-paste(meanTemp$year, meanTemp$monnth)
#plot(meanTemp$mean_temp, pch=16, ylim=c(0, 15))

#meanTemp$mean_temp[is.nan(meanTemp$mean_temp)]<-NA
#plot(meanTemp$mean_temp)

temp$year<- strftime(temp$date, "%Y")
temp$year<-as.integer(temp$year)
temp$month<- strftime(temp$date, "%m")
temp$month<-as.integer(temp$month)

meanTemp<-aggregate(temp$degree_Celsius, by=list(temp$month, temp$year), mean, na.rm=T)
colnames(meanTemp)<-c("month","year","temp")
#Plot different month

lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

p<-list()
rel<-data.frame(month=1:12)
for (i in 1:12){
    df=meanTemp[meanTemp$month==i,]
    result<-lm(year~temp, data=df)
    rel$p.value[i]<-lmp(result)
    rel$r[i]<-summary(result)$r.squared
    p[[i]]<-ggplot(df, aes(x=year, y=temp))+
        geom_point(color="deeppink")+ggtitle(paste0("Month ", i))+
        annotate("text", x=2000, y=,max(df$temp)-0.3, hjust=0, label=paste0("P-value=",round(rel$p.value[i], digits=4)), size=1)+
        annotate("text", x=2000, y=,max(df$temp), hjust=0, label=paste0("R^2=",round(rel$r[i], digits=4)), size=1)
    
}

{pdf("../Output/BayEnv/AK_temp_overtime.pdf", width = 8, height=8)
do.call(grid.arrange, c(p, ncol=2))
dev.off()}
plot(temp$degree_Celsius[temp$month==12])

```

# Standardize the water temperature file
```{r eval=FALSE, message=FALSE, warning=FALSE}
water<-read.csv("../Output/BayEnv/watertemp.csv")
water$C<-(water$mean-32)*5/9
mean<-mean(water$C)
sd<-sd(water$C)
water$C_std<-(water$C-mean)/sd
water2<-data.frame(t(water[,c("pop","C_std")]))
colnames(water2)<-water2[1,]
water2<-water2[,c("BC","CA","PWS","SS","TB","WA")]
#add latitude
lat<-c(52.1,37.69,60.54, 59.93, 58.8, 47.6)
#standardize
lat2<-(lat-mean(lat))/sd(lat)
water2[3,]<-lat2
write.table(water2[2:3,], "../Output/BayEnv/temp_std.txt", sep = "\t",quote = F, row.names = F, col.names = F)

```


# Run BayEnv2 on Farm

*slurm scripts
```{r eval=FALSE, message=FALSE, warning=FALSE}
sink(paste0("../Data/Slurmscripts/bayenv_run.sh"))
cat("#!/bin/bash -l\n")
cat(paste0("#SBATCH --job-name=bayenv \n"))
cat(paste0("#SBATCH --mem=16G \n")) 
cat(paste0("#SBATCH --ntasks=8 \n")) 
cat(paste0("#SBATCH -e =bayenv.err  \n"))
cat(paste0("#SBATCH --time=240:00:00  \n"))
cat(paste0("#SBATCH -p high  \n"))
cat("\n\n")
cat("module load bayenv2 \n\n") 
  
cat("bayenv2 -i /ktist/ph/data/bayenv/Y2017_Pops_snpsfile.txt -p 6 -k 100000 -r 12345 > /ktist/ph/data/bayenv/Y2017_matrix.out \n")
cat("gzip /ktist/ph/data/bayenv/Y2017_matrix.out \n" )
cat("zcat /ktist/ph/data/bayenv/Y2017_matrix.out.gz | perl -lne 'if(/ITER = 100000/){$ok=1}elsif($ok){ print }' > Y2017matrix.txt \n")
sink(NULL)

# Run BayEnv2 with environmental data
sink(paste0("../Data/Slurmscripts/bayenv_run2.sh"))
cat("#!/bin/bash -l\n")
cat(paste0("#SBATCH --job-name=bayenv2 \n"))
cat(paste0("#SBATCH --mem=16G \n")) 
cat(paste0("#SBATCH --ntasks=8 \n")) 
cat(paste0("#SBATCH -e =bayenv2.err  \n"))
cat(paste0("#SBATCH --time=240:00:00  \n"))
cat(paste0("#SBATCH -p high  \n"))
cat("\n\n")
cat("module load bayenv2 \n\n") 
  
cat("bayenv2 -i /ktist/ph/data/bayenv/Y2017_Pops_snpsfile.txt -p 6 -k 100000 -r 12345 > /ktist/ph/data/bayenv/Y2017_matrix.out \n")
cat("gzip /ktist/ph/data/bayenv/Y2017_matrix.out \n" )
cat("zcat /ktist/ph/data/bayenv/Y2017_matrix.out.gz | perl -lne 'if(/ITER = 100000/){$ok=1}elsif($ok){ print }' > Y2017matrix.txt \n")
sink(NULL)




```

## visualize the correlation matrix


```{r eval=FALSE, message=FALSE, warning=FALSE}
library(corrplot)
library(gplots)

# read cov matrices and convert them to correlation matrices
mat17 = as.matrix(read.table(file="../Output/bayenv/Y2017matrix.txt", header=F) )
mat17 = cov2cor(mat17)
rownames(mat17)<-c("BC","CA","PWS","SS","TB","WA")
colnames(mat17)<-c("BC","CA","PWS","SS","TB","WA")
corrplot(mat17,is.corr=T,tl.cex=0.35,order="hclust",title="SNP correlaation")
```

```{r eval=FALSE, message=FALSE, warning=FALSE}

```



```{r eval=FALSE, message=FALSE, warning=FALSE}

```



```{r eval=FALSE, message=FALSE, warning=FALSE}

```



