---
title: "Genes in high Fst Regions"
output:
  html_notebook:
      toc: true 
      toc_float: true
      number_sections: true
      theme: lumen
      highlight: tango
      code_folding: hide
      df_print: paged
---
```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
source("../Rscripts/BaseScripts.R")
library(tidyverse)
library(dplyr)
library(cowplot)
```


# Extract the regions with high Fst from a chromosome  
* Fst is calculated between PCA Groups in PWS
(chr 4, 8,15,20, & the inversion in chr20)

## Create bed files for regions of interest

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Create bed files for regions of interest
chrsize<-read.table("../Data/new_vcf/chr_sizes.bed")
sink("../Output/PCA/genes/chr4_region.bed")
cat("track type=bedGraph \n")
cat(paste0("chr4\t23700000\t",chrsize$V3[chrsize$V1=="chr4"],"\n"))
sink(NULL)

sink("../Output/PCA/genes/chr8_region.bed")
cat("track type=bedGraph \n")
cat(paste0("chr8\t23000000\t",chrsize$V3[chrsize$V1=="chr8"],"\n"))
sink(NULL)

sink("../Output/PCA/genes/chr15_region.bed")
cat("track type=bedGraph \n")
cat(paste0("chr15\t0\t165000000\n"))
sink(NULL)

sink("../Output/PCA/genes/chr20_region.bed")
cat("track type=bedGraph \n")
cat(paste0("chr20\t21600000\t",chrsize$V3[chrsize$V1=="chr20"],"\n"))
sink(NULL)

sink("../Output/PCA/genes/chr20_invregion.bed")
cat("track type=bedGraph \n")
cat(paste0("chr20\t1500000\t3700000\n"))
sink(NULL)
```

## Extract the target region using vcftools and create a new vcf file 

```{bash}
# 2. Create VCF files with the specified regions (using the bed files created above)

cd ~/Projects/PacHerring/
vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/PCA/genes/chr4_region.bed --out Output/PCA/genes/PWS_chr4_region2 --recode --keep-INFO-all

vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/PCA/genes/chr8_region.bed --out Output/PCA/genes/PWS_chr8_region --recode --keep-INFO-all

vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/PCA/genes/chr15_region.bed --out Output/PCA/genes/PWS_chr15_region --recode --keep-INFO-all

vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/PCA/genes/chr20_region.bed --out Output/PCA/genes/PWS_chr20_region --recode --keep-INFO-all

vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/PCA/genes/chr20_invregion.bed --out Output/PCA/genes/PWS_chr20_invregion --recode --keep-INFO-all


### Create VCF with maf00 file to include more snps in the region
vcftools --gzvcf Data/new_vcf/PWS17_MD7000_maf00.vcf.gz --bed Output/PCA/genes/chr8_region.bed --out Output/PCA/genes/PH_chr8_region_maf00 --recode --keep-INFO-all


```

## Create a script to run SnpEff 
```{r echo=TRUE, message=FALSE, warning=FALSE}
#create a bash script to run snpEff
vfiles<-list.files("../Output/PCA/genes/", pattern=".recode.vcf")
sink("~/programs/snpEff/runsnpEff.sh")
cat("#!/bin/bash \n\n")
for (i in 1:length(vfiles)){
    fname<-gsub(".recode.vcf","",vfiles[i])
    cat(paste0("java -Xmx8g -jar snpEff.jar Ch_v2.0.2.99 ~/Projects/PacHerring/Output/PCA/genes/",vfiles[i], " -stats ~/Projects/PacHerring/Output/PCA/genes/",fname,".html >  ~/Projects/PacHerring/Output/PCA/genes/Anno.",fname,".vcf \n"))
    
    #extract the annotation information
    cat(paste0("bcftools query -f '%CHROM %POS %INFO/AF %INFO/ANN\\n' ~/Projects/PacHerring/Output/PCA/genes/Anno.",fname,".vcf > ~/Projects/PacHerring/Output/PCA/genes/",fname,"_annotation \n\n"))

}
sink(NULL)  

## For maf00 file:
sink("~/programs/snpEff/runsnpEff_ch8.sh")
cat("#!/bin/bash \n\n")

cat(paste0("java -Xmx8g -jar snpEff.jar Ch_v2.0.2.99 ~/Projects/PacHerring/Output/PCA/genes/PH_chr8_region_maf00.recode.vcf -stats ~/Projects/PacHerring/Output/PCA/genes/PH_chr8_region_maf00.html >  ~/Projects/PacHerring/Output/PCA/genes/Anno.PH_chr8_region_maf00.vcf \n"))
    
    #extract the annotation information
cat(paste0("bcftools query -f '%CHROM %POS %INFO/AF %INFO/ANN\\n' ~/Projects/PacHerring/Output/PCA/genes/Anno.PH_chr8_region_maf00.vcf > ~/Projects/PacHerring/Output/PCA/genes/Anno.PH_chr8_region_maf00 \n\n"))

sink(NULL)  

```

Run snpEff in its directory
```{bash eval=TRUE, include=FALSE}
cd ~/programs/snpEff
bash runsnpEff/sh
```


## Create summary files of snpEff results (gene annotations in the regions of interest) and reformat as a ShinyGo input 

```{r echo=TRUE, message=FALSE, warning=FALSE}
#create gene list 
gfiles<-list.files("../Output/PCA/genes/", pattern="genes.txt")

for (i in 1:length(gfiles)){
    df<-read.table(paste0("../Output/PCA/genes/",gfiles[i]), sep="\t")
    df<-df[,1:7]
    colnames(df)<-c("GeneName","GeneId","TranscriptId","BioType","variants_impact_HIGH","variants_impact_LOW",	"variants_impact_MODERATE")
    
    fname<-gsub(".genes.txt","",gfiles[i])
    sink(paste0("../Output/PCA/genes/geneIDlist_",fname,".txt"))
    cat(paste0(df$GeneId,"; "))
    sink(NULL)
}

### Convert to NCBI Gene ID using ensemble to use iPath
# http://uswest.ensembl.org/biomart/martview/
#how to video: 
id_c4<-read.csv("../Output/PCA/genes/ch4_inv_mart_export.csv")
id<-id_c4[!is.na(id_c4$NCBI.gene..formerly.Entrezgene..ID),]
sink("../Output/PCA/genes/ch4_inv_geneIDs.txt")
for (i in 1: nrow(id)){
    cat(paste0("NCBI-GI:",id$NCBI.gene..formerly.Entrezgene..ID[i],"\n"))
}
sink(NULL)

id<-read.csv("../Output/PCA/genes/chr15_region_mart_export.txt")
id<-id[!is.na(id$NCBI.gene..formerly.Entrezgene..ID),]
sink("../Output/PCA/genes/ch15_region_geneIDs.txt")
for (i in 1: nrow(id)){
    cat(paste0("NCBI-GI:",id$NCBI.gene..formerly.Entrezgene..ID[i],"\n"))
}
sink(NULL)

```


* Background genes for annotation
```{r}
#create the background gene data for the annotation purpose
df<-read.table("../Data/new_vcf/PWSonly/PWSonly_background.genes.txt", sep="\t")
df<-df[,1:3]
colnames(df)<-c("GeneName","GeneId","TranscriptId")
genes<-unique(df$GeneId)
sink(paste0("../Data/new_vcf/PWSonly/PWSonly_background_geneIDlist.txt"))
cat(paste0(genes,"; "))
sink(NULL)

## Background genes using the maf00 PH file

#Run the following in the snpEff directory
# snpEff_background.sh

#!/bin/bash 

#java -Xmx8g -jar snpEff.jar Ch_v2.0.2.99 /Volumes/Kaho_Data/PacHerring/Data/new_vcf/PH_DP600_7000_minQ20_minMQ30_NS0.5.vcf -stats ~/Projects/PacHerring/Data/new_vcf/PH_background.html >  ~/Projects/PacHerring/Data/new_vcf/background/Anno.PH_background.vcf 


df<-read.table("../Data/new_vcf/background/PH_background.genes.txt", sep="\t")
df<-df[,1:3]
colnames(df)<-c("GeneName","GeneId","TranscriptId")
genes<-unique(df$GeneId)
sink(paste0("../Data/new_vcf/background/PH_background_geneIDlist.txt"))
cat(paste0(genes,"; "))
sink(NULL)

```


<br>

# Results from ShinyGO annotation
## Chr4 [23.7Mb-end] region

![](../Output/PCA/genes/ShinyGO.0.76.2/ch4/barplot.png){width=70%}

## Chr8 [23Mb-end] region
**"NuA4 histone acetyltransferase complex" is involved in double-strand DNA break repair"**

![](../Output/PCA/genes/ShinyGO.0.76.2/ch8/barplot.png){width=70%}

## Chr15 [0-16.5Mb] region

![](../Output/PCA/genes/ShinyGO.0.76.2/ch15/barplot.png){width=70%}

## Chr20 [21.6Mb-end] region

![](../Output/PCA/genes/ShinyGO.0.76.2/ch20_end/barplot.png){width=70%}

## Chr20 [1.5-3.7Mb] inversion region
![](../Output/PCA/genes/ShinyGO.0.76.2/ch20_inv/barplot-2.png){width=80%}


# SOD in the end of Chr20
## Look at AF in Chr20(end)   
- SOD3b is between 25.6 - 26.1Mb
- Each point is a mean AF over 10,000 bp window with 5000 bp steps  

```{r}

groups<-c("Group1","Group2","Group3")
c20af<-data.frame()
for(i in 1:3){
    df<-read.table(paste0("../Data/new_vcf/PWSonly/PWSonly_chr20_",groups[i],"_maf01_freq.frq"),stringsAsFactors = FALSE,header = FALSE, skip=1, col.names = c("chr","pos","n_allele","n_sample","MajorAF","MAF"))
    df$maf<-as.numeric(substr(df$MAF, 3,10))
    df$Maj<-as.numeric(substr(df$MajorAF, 3,10))
    df$Group<-groups[i]
    c20af<-rbind(c20af, df)
}

#Calcualte the mean over 10k window size
library(windowscanr)

Fq<-data.frame()
for(i in 1:3){
    df<-read.table(paste0("../Data/new_vcf/PWSonly/PWSonly_chr20_",groups[i],"_maf01_freq.frq"),stringsAsFactors = FALSE,header = FALSE, skip=1, col.names = c("chr","pos","n_allele","n_sample","MajorAF","MAF"))
    df$maf<-as.numeric(substr(df$MAF, 3,10))
    df$Maj<-as.numeric(substr(df$MajorAF, 3,10))
    freqs_win <- winScan(x = df, position = "pos", values = "maf", win_size = 10000, win_step = 5000, funs = "mean")
    freqs_win <- na.omit(freqs_win)
    freqs_win$Group<-groups[i]
    Fq<-rbind(Fq,freqs_win)
}    

write.csv(Fq,"../Output/PCA/genes/PWSonly_chr20_af_10kwindow.csv")
#Fq<-read.csv("../Output/PCA/genes/PWSonly_chr20_af_10kwindow.csv", row.names = 1)

ggplot(Fq[Fq$win_start>=25000000&Fq$win_end<=27000000,], aes(x=win_mid, y=maf_mean, color=Group))+
    geom_line(size=0.3, alpha=0.9)+
    theme_minimal()+ scale_color_manual(values=cols)+
    ylab("MAF")+xlab('')+theme(legend.title = element_blank())
ggsave("../Output/PCA/genes/PWS_ch20_sod_AF_lineplot.png",width = 7, height = 4, dpi=300)

# Plot the windowed averaged frequenceis 
ggplot(Fq[Fq$win_start>=25000000&Fq$win_end<=27000000,], aes(x=win_mid, y=maf_mean, color=Group))+
    geom_point(size=0.7, alpha=0.4)+
    geom_smooth(span = 0.1,se = FALSE, size=0.5)+
    #geom_line()+
    theme_minimal()+
    scale_color_manual(values=cols)+
    ylab("MAF")+xlab('')+theme(legend.title = element_blank())
ggsave("../Output/PCA/genes/PWS_ch20_sod_AF.png",width = 7, height = 4, dpi=300)
```
![](../Output/PCA/genes/PWS_ch20_sod_AF_lineplot.png){width=80%}

![](../Output/PCA/genes/PWS_ch20_sod_AF.png){width=80%}



## Focus on 26.2-26.5Mb 
- the area with a high allele freq difference
- Plot the actual AF at each loci (not the window average)  
```{r}

## Plot the actual allele frequencies
ggplot(c20af[c20af$pos>=26200000&c20af$pos<=26500000,], aes(x=pos, y=maf, color=Group))+
    geom_point(size=0.5, alpha=0.8)+
    theme_minimal()+
    scale_color_manual(values=cols)+
    ylab("MAF")+xlab('')+theme(legend.title = element_blank())


```

positions near 26,450,000 show the biggest difference.


## Difference in Allele Freq. between Group1 and 3 
- diff(Group3 - Group1) > 0.5

```{r}

# Find the large allele freq differences between Group1 vs. Group3 

diff<-c20af[c20af$Group=="Group1", c("chr","pos","maf")]
diff$diff<-c20af$maf[c20af$Group=="Group3"]-c20af$maf[c20af$Group=="Group1"]

ggplot(diff[abs(diff$diff)>0.5,], aes(x=pos, y=diff))+
    geom_point(size=0.5, alpha=0.8, color=blu)+
    theme_bw()+
    scale_color_manual(values=cols)+
    scale_x_continuous(labels = scales::comma)+
    ylab("AF difference")+xlab('')+theme(legend.title = element_blank())
ggsave("../Output/PCA/genes/PWSonly_chr20_bigDiffAFsites_Group3-Group1.png", width = 6, height = 4, dpi=300)

```

![](../Output/PCA/genes/PWSonly_chr20_bigDiffAFsites_Group3-Group1.png){width=70%}

Differences are aggregated in 24Mb - 26.5Mb 

## Genes in the region:
```{r}
df<-read.table(paste0("../Output/PCA/genes/PWS_chr20_region_annotation"), header = F)
annotations<-data.frame()
    for (j in 1: nrow(df)){
        anns<-unlist(strsplit(df$V4[j], "\\|"))
        anns<-anns[c(2,3,4,5,8,17,18,19,20,23)]
        annotations<-rbind(annotations, anns)
    }     
annotations<-cbind(df[,1:2], annotations)
colnames(annotations)<-c("chr", "pos", "Annotation","Putative_impact","Gene_name", "Gene_ID", "Transcript_biotype","Annotation2","Putative_impact2","Gene_name2", "Gene_ID2", "Transcript_biotype2")


ano<-annotations[!(duplicated(annotations[,3:6])),]
focus<-ano[ano$pos>=26425000& ano$pos<=26490000,]
write.csv(focus, "../Output/PCA/genes/Chr20_genes_in_26.4-26.6M.csv")
library(knitr)
kable(focus, caption = "Genes in the focal region")
```

<br>
<br>

# Find genes in high Fst regions between years in PWS
## From the Null model 
```{r}

outliers<-read.csv("../Output/Fst/Fst_nullshuffling_outliers_PWS.csv", row.names = 1)
#create a bed files
sink("../Output/Fst/Fst_outlier_chr25_region.bed")
cat("track type=bedGraph \n")
cat(paste0("chr19\t", outliers$midPos[outliers$CHROM=="chr19"]-25000, "\t", outliers$midPos[outliers$CHROM=="chr19"]+25000 ,"\n"))
cat(paste0("chr25\t", 805000-25000,"\t",845000+250,"\n"))
sink(NULL)
```

```{bash}
cd ~/Projects/PacHerring/
vcftools --gzvcf Data/new_vcf/PWSonly/PWSonly_NS0.5_maf01.vcf.gz --bed Output/Fst/Fst_outlier_chr25_region.bed --out Output/Fst/Fst_outlier_chr25_region --recode --keep-INFO-all

#Run snpEff in its directory
cd ~/projects/snpEff/
java -Xmx8g -jar snpEff.jar Ch_v2.0.2.99 ~/Projects/PacHerring/Output/Fst/Fst_outlier_chr25_region.recode.vcf -stats ~/Projects/PacHerring/Output/Fst/Fst_outlier_region.recode.vcf.html >  ~/Projects/PacHerring/Output/Fst/Fst_outlier_anno_region.recode.vcf

bcftools query -f '%CHROM %POS %INFO/AF %INFO/ANN\n' ~/Projects/PacHerring/Output/Fst/Fst_outlier_anno_region.recode.vcf > ~/Projects/PacHerring/Output/Fst/Fst_outlier_regoin_PWS_annotation 

```

```{r}
re<-read.table("../Output/Fst/Fst_outlier_regoin_PWS_annotation")
annotations<-data.frame()
for (i in 1: nrow(re)){
    anns<-unlist(strsplit(re$V4[i], "\\|"))
    anns<-anns[c(2,3,4,8)]
    annotations<-rbind(annotations, anns)
}      
re<-cbind(re[,1:3], annotations)
colnames(re)<-c("chr","pos","AF","type","impact","gene name","protein coding")

knitr::kable(re)
```

* mylipa: myosin regulatory light chain interacting protein a
https://zfin.org/ZDB-GENE-030131-7831#gene-ontology
"Predicted to have ubiquitin-protein transferase activity. Involved in gastrulation and negative regulation of Wnt signaling pathway. Localizes to cytoplasm. Is expressed in several structures, including axis; midbrain; neural plate; polster; and retina. Orthologous to human MYLIP (myosin regulatory light chain interacting protein)."
Expresssed upto juveniles

```{bash}

```






