---
title: "COV scan 3 populations"
output:
  html_notebook:
      toc: true 
      toc_float: true
      number_sections: true
      theme: lumen
      highlight: tango
      code_folding: hide
      df_print: paged
---

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
source("../Rscripts/BaseScripts.R")
library(tidyverse)
library(dplyr)
library(cowplot)
library(scales)
```


# Find regions with high covariances in each population
* From Temporal Covariance analysis  -output covariances for each time period

```{r eval=FALSE, message=FALSE, warning=FALSE}

#Find the regions with a high temporal covariance 
pops<-c("PWS","TB","SS")
winsize<-c("50k","75k","100k")
evens<-paste0("chr",seq(2,26, by=2))
cov.list<-list()
covs_all<-list()
k=1
for (p in 1: length(pops)){
    pop<-pops[p]
    for (i in 1: length(winsize)){
        iv<-read.csv(paste0("~/Projects/Pacherring_Vincent/MD7000/3pops_intervals_",winsize[i],"window.csv"), row.names = 1)
        if (p==3) {
            cov23<-read.csv(paste0("~/Projects/Pacherring_Vincent/MD7000/",pop,"_cov23_2017-2006_2006-1996_3Pops_",winsize[i],"window.csv"), header = F)
            covs<-cbind(iv, cov23)
            colnames(covs)[4]<-c("cov23")
            covs$index=1:nrow(covs)
            covs$color<-"col1"
            covs$color[covs$chrom %in% evens]<-"col2"
    
            covs[sapply(covs, is.infinite)] <- NA
            covs[sapply(covs, is.nan)] <- NA
            
            cov.list[[k]]<-covs
            names(cov.list)[k]<-paste0(pop,"_",winsize[i])    
            k=k+1
        }
        else {
            cov12<-read.csv(paste0("~/Projects/Pacherring_Vincent/MD7000/",pop,"_cov12_1996-1991_2006-1996_3Pops_",winsize[i],"window.csv"), header = F)
            cov23<-read.csv(paste0("~/Projects/Pacherring_Vincent/MD7000/",pop,"_cov23_2017-2006_2006-1996_3Pops_",winsize[i],"window.csv"), header = F)
            cov13<-read.csv(paste0("~/Projects/Pacherring_Vincent/MD7000/",pop,"_cov13_2017-2006_1996-1991_3Pops_",winsize[i],"window.csv"), header = F)
            covs<-cbind(iv, cov12, cov23,cov13)
            colnames(covs)[4:6]<-c("cov12","cov23","cov13")
            covs$index=1:nrow(covs)
    
            covs$color<-"col1"
            covs$color[covs$chrom %in% evens]<-"col2"
    
            covs[sapply(covs, is.infinite)] <- NA
            covs[sapply(covs, is.nan)] <- NA
            
            cov.list[[k]]<-covs
            names(cov.list)[k]<-paste0(pop,"_",winsize[i])    
            k=k+1
        }
    }
}
```



# Find the covariance lower cut off values  
```{r eval=FALSE, message=FALSE, warning=FALSE}

cv<-c("cov12","cov13","cov23")
winsize<-c("50k","75k","100k")

cvrange<-data.frame(pop=c(paste0(pops[1:2],"_", cv[1]),paste0(pops[1:2],"_", cv[2]),paste0(pops,"_", cv[3])))
k=1
for (i in 1:length(cv)){
    if (i==1|i==2){
        if (i==1) k=1
        if(i==2) k=3
        for (w in 1: length(winsize)){
            #PWS
            df1<-cov.list[[paste0("PWS_", winsize[w])]]
            df1<-df1[order(df1[,cv[i]], decreasing=T),]
            n<-ceiling(nrow(df1)*0.01) #top1% region
            df1$top1<-"N"
            df1$top1[1:n]<-"PWS"
            
            rg<-range(df1[df1$top1=="PWS",cv[i]], na.rm=T)
            cvrange[k,paste0(winsize[w])]<-paste0(rg[1],"-",rg[2])
           
            #tb
            df2<-cov.list[[paste0("TB_", winsize[w])]]
            df2<-df2[order(df2[,cv[i]], decreasing=T),]
            df2$top1<-"N"
            df2$top1[1:n]<-"TB"
            
            rg2<-range(df2[df2$top1=="TB", cv[i]], na.rm=T)
            cvrange[(k+1),paste0(winsize[w])]<-paste0(rg2[1],"-",rg2[2])
 
        }
    }
   
    if (i==3){
        k=5
        for (w in 1: length(winsize)){
        #pws
        df1<-cov.list[[paste0("PWS_", winsize[w])]]
        df1<-df1[,c("chrom","start","end","cov23")]
        df1<-df1[order(df1$cov23, decreasing=T),]
        n<-ceiling(nrow(df1)*0.01) #top1% region
        df1$top1<-"N"
        df1$top1[1:n]<-"PWS"
        
        rg<-range(df1[df1$top1=="PWS",cv[i]], na.rm=T)
        cvrange[k,paste0(winsize[w])]<-paste0(rg[1],"-",rg[2])
           
    
        #tb
        df2<-cov.list[[paste0("TB_", winsize[w])]]
        df2<-df2[,c("chrom","start","end","cov23")]
        df2<-df2[order(df2$cov23, decreasing=T),]
        df2$top1<-"N"
        df2$top1[1:n]<-"TB"
        rg2<-range(df2[df2$top1=="TB", cv[i]], na.rm=T)
        cvrange[(k+1),paste0(winsize[w])]<-paste0(rg2[1],"-",rg2[2])
    
        #ss
        df3<-cov.list[[paste0("SS_", winsize[w])]]
        df3<-df3[,c("chrom","start","end","cov23")]
        df3<-df3[order(df3$cov23, decreasing=T),]
        df3$top1<-"N"
        df3$top1[1:n]<-"SS"
        rg3<-range(df3[df3$top1=="SS", cv[i]], na.rm=T)
        cvrange[(k+2),paste0(winsize[w])]<-paste0(rg3[1],"-",rg3[2])
        }
    }
}

write.csv(cvrange, "../Output/COV/COVscan_3pop/COVranges_check.csv")
cvrange<-read.csv("../Output/COV/COVscan_3pop/COVranges_check.csv", row.names = 1,check.names=FALSE)

cvs<-melt(cvrange, id.vars = "pop")
cvs<-cvs %>%
  separate(value, c("low", "high"), "-")
cvs$low<-as.numeric(cvs$low)
cvs$high<-as.numeric(cvs$high)
cvs<-cvs%>%
  separate(pop, c("pop", "cov"), "_")

ggplot(cvs, aes(x=cov, y=high, fill=pop))+
    facet_wrap(~variable)+
    geom_crossbar(aes(ymin=low, ymax=high), width=0.5, position=position_dodge(width = 1))+
    ylab("Range of covariances")+
    theme_light()+
    geom_vline(xintercept=c(1.5,2.5), color="gray")+
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
ggsave("../Output/COV/COVscan_3pop/TempCov_Range_comparison.png", width = 7, height = 3, dpi=300)

ggplot(cvs, aes(x=cov, y=low, color=pop))+
    facet_wrap(~variable)+
    geom_point()+
    ylab("Lower limit of top 1% covariance")+
    theme_light()+
    geom_vline(xintercept=c(1.5,2.5), color="gray")+
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
ggsave("../Output/COV/COVscan_3pop/TempCov_Range_lowLimit_comparison.png", width = 7, height = 3, dpi=300)

```

![](../Output/COV/COVscan_3pop/TempCov_Range_comparison.png)

![](../Output/COV/COVscan_3pop/TempCov_Range_lowLimit_comparison.png)  

## Find the covariance ranges for different populations to set the lowest cutoff values 
```{r eval=FALSE, message=FALSE, warning=FALSE}

cv<-c("cov12","cov13","cov23")
winsize<-"100k"

#cvrange<-data.frame(pop=c(paste0(pops[1:2],"_", cv[1]),paste0(pops[1:2],"_", cv[2]),paste0(pops,"_", cv[3])))
k=1
w=1
for (c in 1:length(cv)){
    if (c==1|c==2){
        #PWS
        df1<-cov.list[[paste0("PWS_", winsize[w])]]
        df1<-df1[order(df1[,cv[c]], decreasing=T),]
        n<-ceiling(nrow(df1)*0.01) #top1% region
        df1$pop<-"PWS"
        #tb
        df2<-cov.list[[paste0("TB_", winsize[w])]]
        df2<-df2[order(df2[,cv[c]], decreasing=T),]
        df2$pop<-"TB"

        covs<-rbind(df1, df2)
        colnames(covs)[colnames(covs)==cv[c]]<-"cov"
        #remove one extreme
        covs<-covs[covs$cov>-0.3,]
        
        #create a table of numbers of loci in the cov bins
        counts<-data.frame(pop=c("PWS","TB"))
        bins<-seq(-0.05, 0.05,0.01)
        for (i in 1: length(bins)){
            if (i==1){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov<bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov<bins[i]&covs$pop=="TB"])
            }
            if (i>1&i<11){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i-1]&covs$cov<bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i-1]&covs$cov<bins[i]&covs$pop=="TB"])
        
            }
            if(i==11){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i]&covs$pop=="TB"])
            }
        }
        
        counts2<-data.frame(t(counts))
        colnames(counts2)<-counts2[1,]
        counts2<-counts2[-1,]
        counts2$bin<-gsub("bin_","",rownames(counts2))
        countm<-melt(counts2, id.vars="bin")
        countm$value<-as.integer(countm$value)
        countm$bin<-as.numeric(countm$bin)
        
        ggplot(countm, aes(x=as.factor(bin), y=value, fill=variable))+
            geom_bar(position=position_dodge(width = 0.8), stat="identity", width = 0.8)+
            scale_fill_manual(values=cols)+
            xlab("Covariance")+ylab("Count")+
            theme(legend.title = element_blank())
        ggsave(paste0("../Output/COV/COVscan_3pop/",cv[c],"_coutns_PWS_TB.png"), width = 6,height=4, dpi=300)
            
        ggplot(covs, aes(x=cov, fill=pop))+
            geom_histogram(data=subset(covs, pop=="PWS"),color="gray60", alpha = 0.5,fill=cols[2], bins=100)+
            geom_histogram(data=subset(covs, pop=="TB"), color="gray60", alpha = 0.5,fill=cols[1], bins=100)+
            geom_vline(xintercept = mean(covs$cov[covs$pop=="PWS"], na.rm=T), color=cols[2])+
            geom_vline(xintercept = mean(covs$cov[covs$pop=="TB"], na.rm=T), color=cols[1])+
            xlim(-0.1,0.1)+
            annotate("rect", xmin=0.05, xmax=0.053,ymax=700,ymin=680, alpha=0.8, fill=cols[2])+
            annotate("rect", xmin=0.05, xmax=0.053,ymax=660,ymin=640, alpha=0.8, fill=cols[1])+
            geom_text(label="PWS", x= 0.058, y=700, size=3.5, color="gray20",vjust=1,hjust=0)+
            geom_text(label="TB",  x= 0.058, y=660, size=3.5, color="gray20",vjust=1, hjust=0)
        ggsave(paste0("../Output/COV/COVscan_3pop/",cv[c],"_histgram_PWS_TB.png"), width = 6,height=4, dpi=300)
        }
   
    if (c==3){
        #PWS
        df1<-cov.list[[paste0("PWS_", winsize[w])]]
        df1<-df1[,c("chrom","start","end","cov23")]
        df1$pop<-"PWS"
        #tb
        df2<-cov.list[[paste0("TB_", winsize[w])]]
        df2<-df2[,c("chrom","start","end","cov23")]
        df2$pop<-"TB"
        #ss
        df3<-cov.list[[paste0("SS_", winsize[w])]]
        df3<-df3[,c("chrom","start","end","cov23")]
        df3$pop<-"SS"
        
        covs<-rbind(df1,df2,df3)
        colnames(covs)[colnames(covs)==cv[c]]<-"cov"
        #remove one extreme
        covs<-covs[covs$cov>-0.3,]
        
        #create a table of numbers of loci in the cov bins
        counts<-data.frame(pop=pops)
        bins<-seq(-0.05, 0.05,0.01)
        for (i in 1: length(bins)){
            if (i==1){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov<bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov<bins[i]&covs$pop=="TB"])
                counts[3,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov<bins[i]&covs$pop=="SS"])
            }
            if (i>1&i<11){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i-1]&covs$cov<bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i-1]&covs$cov<bins[i]&covs$pop=="TB"])
                counts[3,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i-1]&covs$cov<bins[i]&covs$pop=="SS"])
            }
            if(i==11){
                counts[1,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i]&covs$pop=="PWS"])
                counts[2,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i]&covs$pop=="TB"])
                counts[3,paste0("bin_",bins[i])]<-length(covs$cov[covs$cov>=bins[i]&covs$pop=="SS"])
            }
        }
        
        counts2<-data.frame(t(counts))
        colnames(counts2)<-counts2[1,]
        counts2<-counts2[-1,]
        counts2$bin<-gsub("bin_","",rownames(counts2))
        countm<-melt(counts2, id.vars="bin")
        countm$value<-as.integer(countm$value)
        countm$bin<-as.numeric(countm$bin)
        
        ggplot(countm, aes(x=as.factor(bin), y=value, fill=variable))+
            geom_bar(position=position_dodge(width = 0.8), stat="identity", width = 0.8)+
            scale_fill_manual(values=cols)+
            xlab("Covariance")+ylab("Count")+
            theme(legend.title = element_blank())
        ggsave(paste0("../Output/COV/COVscan_3pop/",cv[c],"_coutns_3pops.png"), width = 6,height=4, dpi=300)
        
        plots<-list()    
        for (j in 1:3){
            if (j==1) color_name<-cols[2]
            if (j==2) color_name<-cols[1]
            if (j==3) color_name<-cols[3]
            
            plots[[j]]<- ggplot(covs, aes(x=cov, fill=pop))+
                    geom_histogram(data=subset(covs, pop==pops[j]),color="gray60", alpha = 0.5,fill=color_name, bins=100)+
                    geom_vline(xintercept = mean(covs$cov[covs$pop==pops[j]], na.rm=T), color=color_name)+
                    xlim(-0.1,0.1)+ggtitle(pops[j])
        }
        {png(paste0("../Output/COV/COVscan_3pop/",cv[c],"_histgram_3pops.png"), height = 8, width = 4.6, res=300, units = "in")
        do.call(grid.arrange, c(plots, ncol=1))
        dev.off()}
        }
}

```

![](../Output/COV/COVscan_3pop/COV12_coutns_PWS_TB.png){width=65%}  
![](../Output/COV/COVscan_3pop/COV12_histgram_PWS_TB.png){width=65%}


## Use the lowest covariance value for each window and each time period  -not using it
```{r eval=FALSE, message=FALSE, warning=FALSE}
lows<-aggregate(cvs$low, by=list(cvs$variable, cvs$cov), min)
colnames(lows)<-c("window","cov","low")

lows
# window   cov        low
#1    50k cov12 0.03981759
#2    75k cov12 0.03292430
#3   100k cov12 0.02874841
#4    50k cov13 0.03973678
#5    75k cov13 0.03541218
#6   100k cov13 0.03102712
#7    50k cov23 0.04554084
#8    75k cov23 0.03672019
#9   100k cov23 0.03246524


# Outliers based on the new low cut-off values for each windows. 
cov12<-data.frame()
cov23<-data.frame()
cov13<-data.frame()

for (i in 1:length(cov.list)){
 if (grepl("PWS",names(cov.list)[i])|grepl("TB",names(cov.list)[i]))
     {
    covs<-cov.list[[i]]
    
    pop<-gsub("_.+",'', names(cov.list)[i])
    win<-gsub(paste0(pop,"_"), '', names(cov.list)[i])
    
    #outlier cutoff value
    x<-lows$low[lows$window==win&lows$cov=="cov12"]
    covs12_top<-subset(covs, cov12>=x)
    covs12_top<-covs12_top[order(covs12_top$chrom, covs12_top$start),]
    covs12_top$window<-win
    covs12_top$pop<-pop
    cov12<-rbind(cov12, covs12_top)
    
    covs<-covs[order(covs$cov13, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov13"]
    covs13_top<-subset(covs, cov13>=x)
    covs13_top<-covs13_top[order(covs13_top$chrom, covs13_top$start),]
    covs13_top$window<-win
    covs13_top$pop<-pop
    cov13<-rbind(cov13, covs13_top)
    
    covs<-covs[order(covs$cov23, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov23"]
    covs23_top<-subset(covs[,c("chrom","start","end","cov23","index","color")], cov23>=x)
    covs23_top<-covs23_top[order(covs23_top$chrom, covs23_top$start),]
    covs23_top$window<-win
    covs23_top$pop<-pop
    cov23<-rbind(cov23, covs23_top)
 }
 if (grepl("SS",names(cov.list)[i])){
    covs<-cov.list[[i]]
    
    pop<-gsub("_.+",'', names(cov.list)[i])
    win<-gsub(paste0(pop,"_"), '', names(cov.list)[i])
    
    covs<-covs[order(covs$cov23, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov23"]
    covs23_top<-subset(covs, cov23>=x)
    covs23_top<-covs23_top[order(covs23_top$chrom, covs23_top$start),]
    covs23_top$window<-win
    covs23_top$pop<-pop
    cov23<-rbind(cov23, covs23_top)
    }
}

write.csv(cov12, "../Output/COV/COVscan_3pop/3pops_top1percent_outlier_cutoff.cov12.csv",row.names = F)
write.csv(cov23, "../Output/COV/COVscan_3pop/3pops_top1percent_outlier_cutoff.cov23.csv",row.names = F)
write.csv(cov13, "../Output/COV/COVscan_3pop/3pops_top1percent_outlier_cutoff.cov13.csv",row.names = F)
```


## Use the PWS's lowest covariance value for TB and SS 

```{r eval=FALSE, message=FALSE, warning=FALSE}
lows<-cvs[cvs$pop=="PWS",]
names(lows)[3]<-"window"
#low for PWS 100k
#   pop   cov variable        low      high
#15 PWS cov12     100k 0.02874841 0.0821782
#17 PWS cov13     100k 0.03102712 0.1036491
#19 PWS cov23     100k 0.03274974 0.2022322


# Outliers based on the new low cut-off values 100k window. 
cov12<-data.frame()
cov23<-data.frame()
cov13<-data.frame()

for (i in 1:length(cov.list)){
 if (grepl("PWS",names(cov.list)[i])|grepl("TB",names(cov.list)[i])){
    covs<-cov.list[[i]]
    
    pop<-gsub("_.+",'', names(cov.list)[i])
    win<-gsub(paste0(pop,"_"), '', names(cov.list)[i])
    
    #outlier cutoff value
    x<-lows$low[lows$window==win&lows$cov=="cov12"]
    covs12_top<-subset(covs, cov12>=x)
    covs12_top<-covs12_top[order(covs12_top$chrom, covs12_top$start),]
    covs12_top$window<-win
    covs12_top$pop<-pop
    cov12<-rbind(cov12, covs12_top)
    
    covs<-covs[order(covs$cov13, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov13"]
    covs13_top<-subset(covs, cov13>=x)
    covs13_top<-covs13_top[order(covs13_top$chrom, covs13_top$start),]
    covs13_top$window<-win
    covs13_top$pop<-pop
    cov13<-rbind(cov13, covs13_top)
    
    covs<-covs[order(covs$cov23, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov23"]
    covs23_top<-subset(covs[,c("chrom","start","end","cov23","index","color")], cov23>=x)
    covs23_top<-covs23_top[order(covs23_top$chrom, covs23_top$start),]
    covs23_top$window<-win
    covs23_top$pop<-pop
    cov23<-rbind(cov23, covs23_top)
 }
 if (grepl("SS",names(cov.list)[i])){
    covs<-cov.list[[i]]
    
    pop<-gsub("_.+",'', names(cov.list)[i])
    win<-gsub(paste0(pop,"_"), '', names(cov.list)[i])
    
    covs<-covs[order(covs$cov23, decreasing=T),]
    x<-lows$low[lows$window==win&lows$cov=="cov23"]
    covs23_top<-subset(covs, cov23>=x)
    covs23_top<-covs23_top[order(covs23_top$chrom, covs23_top$start),]
    covs23_top$window<-win
    covs23_top$pop<-pop
    cov23<-rbind(cov23, covs23_top)
    }
}

write.csv(cov12, "../Output/COV/COVscan_3pop/cutoff/3pops_top1percent_outlier_cutoffPWS.cov12.csv",row.names = F)
write.csv(cov23, "../Output/COV/COVscan_3pop/cutoff/3pops_top1percent_outlier_cutoffPWS.cov23.csv",row.names = F)
write.csv(cov13, "../Output/COV/COVscan_3pop/cutoff/3pops_top1percent_outlier_cutoffPWS.cov13.csv",row.names = F)
```



# Find overlapping outlier regions between different window sizes (with cutoff to PWS lows) 
```{r eval=FALSE, message=FALSE, warning=FALSE}
# How window size affects outlier regions
#COV12
cv<-c("cov12","cov13","cov23")
winsize<-c("50k","75k","100k")
pairs<-t(combn(winsize, 2))
pairs<-data.frame(pairs)
colnames(pairs)<-paste0("window",1:2)
Ov1<-pairs
Ov2<-pairs
Ov3<-pairs
populations<-c("PWS","TB")
rm(ss)
for (i in 1:length(cv)){
    df<-read.csv(paste0("../Output/COV/COVscan_3pop/cutoff/3pops_top1percent_outlier_cutoffPWS.", cv[i], ".csv"))
    df$id<-paste0(df$chrom,"_",df$start)
    pws<-df[df$pop=="PWS",]
    tb<-df[df$pop=="TB",]
    if (i==3) ss<-df[df$pop=="SS",]
        
    for(j in 1:nrow(pairs)){
        isec<-intersect(pws$id[pws$window==pairs[j,1]], pws$id[pws$window==pairs[j,2]]) 
        Ov1[j, cv[i]]<-length(isec)
        isec2<-intersect(tb$id[tb$window==pairs[j,1]], tb$id[tb$window==pairs[j,2]]) 
        Ov2[j, cv[i]]<-length(isec2)
        if (i==3) {
              isec3<-intersect(ss$id[ss$window==pairs[j,1]], ss$id[ss$window==pairs[j,2]]) 
              Ov3[j, cv[i]]<-length(isec3)
         }
        #### Check chromosome region overlap +-200,000 bases
        
        for(p in 1:length(pops)){
            if (p==3 &!exists("ss")) next
            else {
                if (p==1) df_pop<-pws
            if (p==2) df_pop<-tb
            if (p==3 & exists("ss")) df_pop<-ss
            if (p==3 &!exists("ss")) next
            
            time1<-df_pop[df_pop$window==pairs[j,1],]
            time2<-df_pop[df_pop$window==pairs[j,2],]
             
            overlps<-data.frame()
            overlps2<-data.frame()
        
            for (n in 1: nrow(time1)){
                re<-time2[time2$chrom==time1$chrom[n],]
                if (nrow(re)>=1){
                    for (s in 1: nrow(re)){
                        if (re$start[s]<=time1$start[n]+200000 & re$start[s]>=time1$start[n]-200000){
                            overlps<-rbind(overlps, re[s,])
                            overlps2<-rbind(overlps2,time1[n,])}
                }}
            }      
            #Overlapping windows:
            ov<-data.frame(id=overlps$id)
            for (n in 1: nrow(overlps)){
                if (overlps$start[n]<overlps2$start[n]) {ov$start[n]<-overlps$start[n]; ov$end[n]<-overlps2$end[n]}
                if (overlps$start[n]>=overlps2$start[n]) {ov$start[n]<-overlps2$start[n];ov$end[n]<-overlps$end[n]}
            }
            ov[,paste0("cov.",pairs[j,1])]<-overlps[,cv[i]]
            ov[,paste0("cov.",pairs[j,2])]<-overlps2[,cv[i]]
            write.csv(ov, paste0("../Output/COV/COVscan_3pop/Overlap_regions_cutoff_",pops[p],".", pairs[j,1],".",pairs[j,2], ".",cv[i], "_plusminus200k.csv"), row.names = F)
                
            }
            
        }
    }
}
#write.csv(Ov1, paste0("../Output/COV/COVscan_3pop/Overlapping_window_counts_PWS.csv"))
#write.csv(Ov2, paste0("../Output/COV/COVscan_3pop/Overlapping_window_counts_TB.csv"))
#write.csv(Ov3, paste0("../Output/COV/COVscan_3pop/Overlapping_window_counts_SS.csv"))
```





## Overlapping outlier regions between different populations 
```{r eval=FALSE, message=FALSE, warning=FALSE}

cv<-c("cov12","cov13","cov23")
pairs<-t(combn(pops, 2))
pairs<-data.frame(pairs)
colnames(pairs)<-paste0("pop",1:2)
Ov_direct<-data.frame(cov=c(cv[1:2],"cov23-PT","cov23-PS","cov23-ST" ,"cov23-3"))
Ov_300<-data.frame(cov=c(cv[1:2],"cov23-PT","cov23-PS","cov23-ST" ,"cov23-3"))
for (i in 1:length(cv)){
    df<-read.csv(paste0("../Output/COV/COVscan_3pop/3pops_top1percent_outlier_regions.", cv[i], "_new.csv"))
    df<-df[df$window=="100k",]
    df$id<-paste0(df$chrom,"_",df$start)
    
    if (i!=3){
        #exact overlaps
        isec<-intersect(df$id[df$pop=="PWS"], df$id[df$pop=="TB"]) 
        Ov_direct$count[i]<-length(isec)
        
        #### Check chromosome region overlap +-200,000 bases
        pop1<-df[df$pop=="PWS",]
        pop2<-df[df$pop=="TB",]
        overlps<-data.frame()
        overlps2<-data.frame()
        for (n in 1: nrow(pop1)){
            re<-pop2[pop2$chrom==pop1$chrom[n],]
            if (nrow(re)>=1){
                for (s in 1: nrow(re)){
                    if (re$start[s]<=pop1$start[n]+300000 & re$start[s]>=pop1$start[n]-300000){
                        overlps<-rbind(overlps, re[s,])
                        overlps2<-rbind(overlps2,pop1[n,])}
                }
            }
        }
        # Merge two tables into one summary overlap table:
        ov<-data.frame(id=overlps$id)
        for (n in 1: nrow(overlps)){
            if (overlps$start[n]<overlps2$start[n]) {ov$start[n]<-overlps$start[n]; ov$end[n]<-overlps2$end[n]}
            if (overlps$start[n]>=overlps2$start[n]) {ov$start[n]<-overlps2$start[n];ov$end[n]<-overlps$end[n]}
        }
        ov[,"cov.PWS"]<-overlps[,4]
        ov[,"cov.TB"]<-overlps2[,4]
        write.csv(ov, paste0("../Output/COV/COVscan_3pop/Overlap_regions_",cv[i],"_plusminus300k.csv"), row.names = F)
        Ov_300$count[i]<-nrow(ov)
        }
        
    if (i==3){
        isec<-intersect(df$id[df$pop=="PWS"], df$id[df$pop=="TB"]) 
        isec2<-intersect(df$id[df$pop=="PWS"], df$id[df$pop=="SS"]) 
        isec3<-intersect(df$id[df$pop=="SS"], df$id[df$pop=="TB"]) 
        Ov_direct$count[i]<-length(isec)
        Ov_direct$count[i+1]<-length(isec2)
        Ov_direct$count[i+2]<-length(isec)
        Ov_direct$count[i+3]<-length(intersect(df$id[df$pop=="SS"], intersect(df$id[df$pop=="PWS"], df$id[df$pop=="TB"])))
        
        for(j in 1:nrow(pairs)){
        #### Check chromosome region overlap +-200,000 bases
            pop1<-df[df$pop==pairs[j,1],]
            pop2<-df[df$pop==pairs[j,2],]
            overlps<-data.frame()
            overlps2<-data.frame()
            for (n in 1: nrow(pop1)){
                re<-pop2[pop2$chrom==pop1$chrom[n],]
                if (nrow(re)>=1){
                    for (s in 1: nrow(re)){
                        if (re$start[s]<=pop1$start[n]+300000 & re$start[s]>=pop1$start[n]-300000){
                            overlps<-rbind(overlps, re[s,])
                            overlps2<-rbind(overlps2,pop1[n,])}
                    }
                }
            }
        # Merge two tables into one summary overlap table:
            ov<-data.frame(id=overlps$id)
            for (n in 1: nrow(overlps)){
                if (overlps$start[n]<overlps2$start[n]) {ov$start[n]<-overlps$start[n]; ov$end[n]<-overlps2$end[n]}
                if (overlps$start[n]>=overlps2$start[n]) {ov$start[n]<-overlps2$start[n];ov$end[n]<-overlps$end[n]}
            }
        
            ov[,paste0("cov.",pairs[j,1])]<-overlps[,4]
            ov[,paste0("cov.",pairs[j,2])]<-overlps2[,4]
            ov<-ov[!duplicated(ov),]
            write.csv(ov, paste0("../Output/COV/COVscan_3pop/Overlap_regions_",cv[i],"_",pairs[j,1],".", pairs[j,2],"_plusminus300k.csv"), row.names = F)
            Ov_300$count[i+j-1]<-nrow(ov)
    }
    }
}
write.csv(Ov_direct, paste0("../Output/COV/COVscan_3pop/DDirect_Overlapping_regions_counts_3pop_summary.csv"))
Ov_300$count[6]<-NA
write.csv(Ov_300, paste0("../Output/COV/COVscan_3pop/Overlapping_regions_counts_3pop_plusMinus300k.csv"))

```


# Run the snpEff pipeline to find annotation in the outlier regions (50k, 75k, 100k window +-100k) -all inclusive 
## Create a script to run SnpEff 

Create VCF files with selected regions & run snpEff  
```{r echo=TRUE, message=FALSE, warning=FALSE}
#Create bed files
cv<-c("cov12","cov13","cov23")
#Prevent scientific notation in bed files
options(scipen=999)

#The first line of bed files is often not red by vcftools
#for each window and together
for (i in 1:3){
    df<-read.csv(paste0("../Output/COV/COVscan_3pop/3pops_top1percent_outlier_cutoff.",cv[i],".csv"))
    
    df$start<-df$start-100000
    df$end<-df$end+100000
    dfp<-df[df$pop=="PWS",1:3]
    dfp<-dfp[order(dfp$chrom),]
    colnames(dfp)<-c('track type=bedGraph', '1','1')
    write.csv(dfp, paste0("../Output/COV/COVscan_3pop/PWS_outliers_all.",cv[i],".csv"),row.names = F, col.names = T)
    dfp50<-df[df$window=="50k"& df$pop=="PWS",1:3]
    colnames(dfp50)<-c('track type=bedGraph', '1','1')
    write.table(dfp50, paste0("../Output/COV/COVscan_3pop/cutoff/PWS_outliers_50k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
     dfp75<-df[df$window=="75k"& df$pop=="PWS",1:3]
    colnames(dfp75)<-c('track type=bedGraph', '1','1')
    write.table(dfp75, paste0("../Output/COV/COVscan_3pop/cutoff/PWS_outliers_75k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
     dfp100<-df[df$window=="100k"& df$pop=="PWS",1:3]
    colnames(dfp100)<-c('track type=bedGraph', '1','1')
    write.table(dfp100, paste0("../Output/COV/COVscan_3pop/cutoff/PWS_outliers_100k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
    
    dft<-df[df$pop=="TB",1:3]
    dft<-dft[order(dft$chrom),]
    colnames(dft)<-c('track type=bedGraph', '1','1')
    write.csv(dft, paste0("../Output/COV/COVscan_3pop/TB_outliers_all.",cv[i],".csv"),row.names = F, col.names = F)
    dft50<-df[df$window=="50k"& df$pop=="TB",1:3]
    colnames(dft50)<-c('track type=bedGraph', '1','1')
    write.table(dft50, paste0("../Output/COV/COVscan_3pop/cutoff/TB_outliers_50k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
     dft75<-df[df$window=="75k"& df$pop=="TB",1:3]
    colnames(dft75)<-c('track type=bedGraph', '1','1')
    write.table(dft75, paste0("../Output/COV/COVscan_3pop/cutoff/TB_outliers_75k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
     dft100<-df[df$window=="100k"& df$pop=="TB",1:3]
    colnames(dft100)<-c('track type=bedGraph', '1','1')
    write.table(dft100, paste0("../Output/COV/COVscan_3pop/cutoff/TB_outliers_100k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
    
    
    
    if (i==3){
        dfs<-df[df$pop=="SS",1:3]
        dfs<-dfs[order(dfs$chrom),]
        colnames(dfs)<-c('track type=bedGraph', '1','1')
        write.csv(dfs, paste0("../Output/COV/COVscan_3pop/SS_outliers_all.",cv[i],"_.csv"),row.names = F, col.names = F)
            dfs50<-df[df$window=="50k"& df$pop=="SS",1:3]
        colnames(dfs50)<-c('track type=bedGraph', '1','1')
        write.table(dfs50, paste0("../Output/COV/COVscan_3pop/cutoff/SS_outliers_50k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
        dfs75<-df[df$window=="75k"& df$pop=="SS",1:3]
        colnames(dfs75)<-c('track type=bedGraph', '1','1')
        write.table(dfs75, paste0("../Output/COV/COVscan_3pop/cutoff/SS_outliers_75k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
        dfs100<-df[df$window=="100k"& df$pop=="SS",1:3]
        colnames(dfs100)<-c('track type=bedGraph', '1','1')
        write.table(dfs100, paste0("../Output/COV/COVscan_3pop/cutoff/SS_outliers_100k",cv[i],".bed"),quote = F, row.names = F, col.names = T,sep = "\t")
        
    }
}

#clean up the bed files
bfiles<-list.files("../Output/COV/COVscan_3pop/", pattern="_outliers_all.+.csv$")

for (i in 1:length(bfiles)){
     df1<-read.csv(paste0("../Output/COV/COVscan_3pop/",bfiles[i]))
     chroms<-unique(df1$track.type.bedGraph)
     df1.new<-data.frame()
     
     # Within each chromosome
     for (c in 1: length(chroms)){
        dfc<-df1[df1$track.type.bedGraph==chroms[c],]
        dfc<-dfc[order(dfc$X1),]
        k=1
        n=(nrow(dfc)-1)
        while (k<n){
            if (!(dfc$X1[(k+1)] %in% c(dfc$X1[k],dfc$X1.1[k]))) {df1.new<-rbind(df1.new, dfc[k,]); k=k+1}
            if (dfc$X1[(k+1)] %in% c(dfc$X1[k],dfc$X1.1[k])) {
                        newrow<-data.frame(track.type.bedGraph=chroms[c],X1=dfc$X1[n],X1.1=dfc$X1.1[n+1])
                        df1.new<-rbind(df1.new, newrow)
                        k=k+2
            }
        }
       
        if (!(dfc$X1[n+1] %in% c(dfc$X1[n],dfc$X1.1[n]))) {df1.new<-rbind(df1.new, dfc[n:(n+1),])}
        if (dfc$X1[(n+1)] %in% c(dfc$X1[n],dfc$X1.1[n])) {
                    newrow<-data.frame(track.type.bedGraph=chroms[c],X1=dfc$X1[n],X1.1=dfc$X1.1[n+1])
                    df1.new<-rbind(df1.new, newrow)
        }
        
     }
     
    bedname<-gsub("csv",'bed',bfiles[i])
    write.table(df1.new, paste0("../Output/COV/COVscan_3pop/cutoff/", bedname),quote = F, row.names = F, col.names = T,sep = "\t")
}
    
 
# Create a bash script to run snpEff
bedfiles<-list.files("../Output/COV/COVscan_3pop/cutoff/", pattern="*.bed")

sink("../COVscan_createVCFs_3Pops_cutoff.sh")
cat("#!/bin/bash \n\n")
for (i in 1:length(bedfiles)){
    fname<-gsub(".bed",'', bedfiles[i])
    cat(paste0("vcftools --gzvcf Data/new_vcf/3pop/3pops.MD7000_NS0.5_maf05.vcf.gz --bed Output/COV/COVscan_3pop/cutoff/", bedfiles[i], " --out Output/COV/COVscan_3pop/cutoff/", fname," --recode --keep-INFO-all \n"))
}
sink(NULL)  

#create a bash script to run snpEff
vfiles<-list.files("../Output/COV/COVscan_3pop/cutoff/", pattern=".recode.vcf$")

sink("~/programs/snpEff/runsnpEff_cov_3pop_cutoff.sh")
cat("#!/bin/bash \n\n")
for (i in 1:length(vfiles)){
    fname<-gsub(".recode.vcf","",vfiles[i])
    cat(paste0("java -Xmx8g -jar snpEff.jar Ch_v2.0.2.99 ~/Projects/PacHerring/Output/COV/COVscan_3pop/cutoff/",vfiles[i], " -stats ~/Projects/PacHerring/Output/COV/COVscan_3pop/cutoff/",fname,".html >  ~/Projects/PacHerring/Output/COV/COVscan_3pop/cutoff/Anno.",fname,".vcf \n"))
    
    #extract the annotation information
    cat(paste0("bcftools query -f '%CHROM %POS %INFO/AF %INFO/ANN\\n' ~/Projects/PacHerring/Output/COV/COVscan_3pop/cutoff/Anno.",fname,".vcf > ~/Projects/PacHerring/Output/COV/COVscan_3pop/cutoff/",fname,"_annotation \n\n"))

}
sink(NULL)  
```


```{bash eval=FALSE, include=FALSE}
cd ~/Projects/PacHerring
bash COVscan_createVCFs_3pops.sh

cd ~/programs/snpEff
bash runsnpEff_cov_3pop.sh
```




## Create summary gene files from snpEff and run ShinyGO (for all inclusive/cutoff/window-based)

```{r eval=FALSE, message=FALSE, warning=FALSE}
## Create summary files of snpEff results (gene annotations in the regions of interest) and reformat as a ShinyGo input 

#create gene list 
gfiles<-list.files("../Output/COV/COVscan_3pop/cutoff/", pattern="genes.txt")

for (i in 1:length(gfiles)){
    df<-read.table(paste0("../Output/COV/COVscan_3pop/cutoff/",gfiles[i]), sep="\t")
    df<-df[,1:7]
    colnames(df)<-c("GeneName","GeneId","TranscriptId","BioType","variants_impact_HIGH","variants_impact_LOW",	"variants_impact_MODERATE")
    
    fname<-gsub(".genes.txt","",gfiles[i])
    genes<-unique(df$GeneId)
    sink(paste0("../Output/COV/COVscan_3pop/cutoff/geneIDlist_",fname,".txt"))
    cat(paste0(genes,"; "))
    sink(NULL)
}

sink("../Createdir.sh")
cat("#!bin bash\n")
for (i in 1:length(gfiles)){
    gname<-gfiles[i]
    gname<-gsub('.genes.txt','',gname)
    cat(paste0("mkdir ",gname,"\n"))
}
sink(NULL)
```


```{r eval=FALSE, message=FALSE, warning=FALSE}
#Summarize annotation info from SnpEff
afiles<-list.files("../Output/COV/COVscan_3pop/cutoff/", pattern="_annotation$")
for (c in 1:length(afiles)){
    ano<-read.table(paste0("../Output/COV/COVscan_3pop/cutoff/",afiles[c]), header = F)
    annotations<-data.frame()
    for (i in 1: nrow(ano)){
        anns<-unlist(strsplit(ano$V4[i], "\\,|\\|"))
        annm<-data.frame(matrix(anns,ncol = 16, byrow = TRUE))
        annm<-annm[,c(2,3,4,5,8)]
        colnames(annm)<-c("Effect","Putative_impact","Gene_name","Gene_ID","Feature type")
        annm<-annm[!duplicated(annm), ]
        annm$chr<-ano$V1[i]
        annm$chr<-ano$V1[i]
        annm$pos<-ano$V2[i]
        annm$AF<- ano$V3[i]
        annotations<-rbind(annotations, annm)
    }     
    annotations<-annotations[,c(6:8,1:5)]
    annotations<-annotations[!duplicated(annotations[,1:2]),]
    fname<-afiles[c]
    write.csv(annotations, paste0("../Output/COV/COVscan_3pop/cutoff/Genes_",fname,".csv"), row.names = F)
}

```

# ShinyGo Results
## PWS COV12 
* 50k-window
![](../Output/COV/COVscan_3pop/cutoff/ShinyGo/PWS_outliers_50kcov12/barplot.png)

* 75k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_75kcov12/barplot.png)

* 100k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_100kcov12/barplot.png)

* all windows
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_all.cov12/barplot.png)


## PWS COV13 
* 50k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_50kcov13/barplot.png)
* 75k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_75kcov13/barplot-2.png)

* 100k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_100kcov13/barplot.png)

* all windows
![](../Output/COV/COVscan_3pop/cutoff/shinygo/PWS_outliers_all.cov13/barplot.png)

## PWS COV23
* No enriched terms found in all windows.
* "Anatomical structure morphogenesis" in 100k window and all windows with FDR=.1


## TB COV12 
* 50k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_50kcov12/barplot.png)

* 75k-window 
No enriched terms found 

    - Above FDR=0.1  
        -Cytoskeletal motor activity   
        -Fatty acid biosynthetic process   
        -Regulation of cell-cell adhesion   
        -Aorta development   
        -3-oxoacyl-[acyl-carrier-protein] synthase activity  
        -Neuron maturation  
        -Neuronal ion channel clustering   
        -Clustering of voltage-gated sodium channels   
        -Fatty acid metabolic process   
        -Anatomical structure formation involved in morphogenesis   
        -Fatty acid synthase activity   


* 100k-window
    - No enriched terms / FDR>0.1
        -Animal organ morphogenesis 
        -Tissue development 
        -Sensory system development 
        -Positive regulation of canonical Wnt signaling pathway 
        -3-oxoacyl-[acyl-carrier-protein] synthase activity 
        -Regulation of cell shape 
        -Anatomical structure morphogenesis 
        -Regulation of cell morphogenesis 
        -Eye morphogenesis 
        -Camera-type eye morphogenesis 
        -Lateral line development 
        -Pattern specification process 
        -Dynein complex binding 
        -Neural plate morphogenesis 
        -Digestive tract morphogenesis 
        -Lateral line system development 
        -Septin complex 
        -Septin cytoskeleton 
        -Endodeoxyribonuclease activity, producing 3 -phosphomonoesters 

* all windows
No enriched terms


## TB COV13 
* 50k-window
No enriched terms found
* 75k-window/100k-window (FDR >0.1)
    - Negative regulation of cell growth 
    - Polysaccharide biosynthetic process 

* all windows
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_all.cov13/barplot.png)

## TB COV23
* 50k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_50kcov23/barplot.png)

* 75k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_75kcov23/barplot.png)

* 100k-window
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_100kcov23/barplot.png)

* All-windows
![](../Output/COV/COVscan_3pop/cutoff/shinygo/TB_outliers_all.cov23/barplot.png)

## SS COV23
* No enriched terms found

* 100k-window, FDR>0.1
    - Midbrain-hindbrain boundary morphogenesis 
    - Phosphatidate phosphatase activity 

* All windows, FDR>0.05 (<0.1)
    - Ubiquitin ligase complex 


```{r}
#list genes that show up in overlapping windows

p12<-read.csv("../Output/COV/COVscan_3pop/cutoff/Genes_PWS_outliers_all.cov12_annotation.csv")

p12.2<-p12[!duplicated(p12$Gene_name),]
write.csv(p12.2,"../Output/COV/COVscan_3pop/cutoff/PWS_cov12_gene_list.csv")

```






```{r eval=FALSE, message=FALSE, warning=FALSE}

## did not run yet 11/1/22
# Find the intersecting gene names across populations
gfiles2<-list.files("../Output/COV/COVscan_3pop/", pattern="geneIDlist")
glist<-list()
for (i in 1:length(gfiles)){
    df<-read.table(paste0("../Output/COV/COVscan_3pop/",gfiles2[i]), sep=";")
    df<-t(df)
    df<-gsub(" ","",df)
    df2<-df[!is.na(df)]
    vname<-gsub(".txt","",gfiles2[i],)
    vname<-gsub("geneIDlist_","", vname)
    glist[[i]]<-df2
    names(glist)[i]<-vname
}

times<-c("cov12","cov13","cov23")
common<-list()
common_genes<-data.frame(time=times)
for (i in 1:3){
    tlist<-glist[grep(times[i], names(glist))]
    if (i !=3){
        common[[i]]<-intersect(tlist[[1]], tlist[[2]])
        names(common)[[i]]<-times[i]
        common_genes$PWS[i]<-length(tlist[[grep("PWS", names(tlist))]])
        common_genes$TB[i]<-length(tlist[[grep("TB", names(tlist))]])
        common_genes$SS[i]<-NA
        common_genes$common_PWS.TB[i]<-length(intersect(tlist[[1]], tlist[[2]]))
    }
    if (i==3){
        common_genes$PWS[i]<-length(tlist[[grep("PWS", names(tlist))]])
        common_genes$TB[i]<-length(tlist[[grep("TB", names(tlist))]])
        common_genes$SS[i]<-length(tlist[[grep("SS", names(tlist))]])
        common_genes$common_PWS.TB[i]<-length(intersect(tlist[[1]], tlist[[3]]))
        common_genes$common_PWS.SS[i]<-length(intersect(tlist[[1]], tlist[[2]]))
        common_genes$common_SS.TB[i]<-length(intersect(tlist[[2]], tlist[[3]]))
        common_genes$common3[i]<-length(intersect(tlist[[1]],intersect(tlist[[2]], tlist[[3]])))
        k=i
        common[[k]]<-intersect(tlist[[1]], tlist[[2]])
        names(common)[[k]]<-paste0(times[i],"_PWS.SS")
        k=k+1
        common[[k]]<-intersect(tlist[[1]], tlist[[3]])
        names(common)[[k]]<-paste0(times[i],"_PWS.TB")
        k=k+1
        common[[k]]<-intersect(tlist[[2]], tlist[[3]])
        names(common)[[k]]<-paste0(times[i],"_SS.TB")
        k=k+1
        common[[k]]<-intersect(tlist[[1]],intersect(tlist[[2]], tlist[[3]]))
        names(common)[[k]]<-paste0(times[i],"_3pops")
        }
}
    
write.csv(common_genes, "../Output/COV/COVscan_3pop/Common_genes_3pops.csv")


#What are the overlapping gene names
#aggregate all gene names
Genes<-data.frame()
for (i in 1:length(gfiles)){
    df<-read.table(paste0("../Output/COV/COVscan_3pop/",gfiles[i]), sep="\t")
    df<-df[,1:2]
    colnames(df)<-c("GeneName","GeneId")
    df<-df[!duplicated(df),]
    Genes<-rbind(Genes, df)
    Genes<-Genes[!duplicated(Genes),]
}

for (i in 1: length(common)){
    gids<-common[[i]]
    df<-data.frame(GeneId=gids)
    
    df<-merge(df, Genes, by="GeneId")
    write.csv(df, paste0("../Output/COV/COVscan_3pop/Common_genes_", names(common)[i],".csv"), row.names = F)
}
```



#### Overlapping gene numbers   
```{r echo=FALSE, message=FALSE, warning=FALSE}

# Summary table
common_genes<-read.csv("../Output/COV/COVscan_3pop/Common_genes_3pops.csv", row.names = 1)
knitr::kable(common_genes)

```



```{r eval=FALSE, message=FALSE, warning=FALSE}

## Check chromosome overlap




```



# Compare results from PH_MD7000 (all pops together) and 3Pops-NS0.5 (filter for 3 pops) VCF files  

```{r eval=FALSE, message=FALSE, warning=FALSE}
pws1<-read.csv("../Output/COV/COVscan_3pop/3pops_top1percent_outlier_regions.cov12_new.csv")
pws1<-pws1[pws1$pop=="PWS" & pws1$window=="100k",]
pws2<-read.csv("../Output/COV/3pops_top1percent_outlier_regions.cov12.csv")
pws2<-pws2[pws2$pop=="PWS"&pws2$window=="100k",]

pws1$vcf<-"3Pops"
pws2$vcf<-"PH"
pws<-rbind(pws1[,c("chrom","start","end","cov12","vcf")],pws2[,c("chrom","start","end","cov12","vcf")])
pws$chrom<-factor(pws$chrom, levels=paste0("chr", 1:26))
ggplot(data=pws,aes(x=start/1000000, y=cov12, color=vcf, fill=vcf))+
    facet_wrap(~chrom)+
    geom_point(position=position_dodge(width = 0.7), alpha=0.5)+xlab("Position (Mb)")+
    ggtitle("PWS COV12")
ggsave("../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov12.png", width = 10, height = 8, dpi=300)
```
![](../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov12.png)

```{r eval=FALSE, message=FALSE, warning=FALSE}

pws1<-read.csv("../Output/COV/COVscan_3pop/3pops_top1percent_outlier_regions.cov23_new.csv")
pws1<-pws1[pws1$pop=="PWS" & pws1$window=="100k",]
pws2<-read.csv("../Output/COV/3pops_top1percent_outlier_regions.cov23.csv")
pws2<-pws2[pws2$pop=="PWS"&pws2$window=="100k",]

pws1$vcf<-"PH"
pws2$vcf<-"3Pops"
pws<-rbind(pws1[,c("chrom","start","end","cov23","vcf")],pws2[,c("chrom","start","end","cov23","vcf")])
pws$chrom<-factor(pws$chrom, levels=paste0("chr", 1:26))
ggplot(data=pws,aes(x=start/1000000, y=cov23, color=vcf, fill=vcf))+
    facet_wrap(~chrom)+
    geom_point(position=position_dodge(width = 0.7), alpha=0.5)+xlab("Position (Mb)")+
    ggtitle("PWS COV23")
ggsave("../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov23.png", width = 10, height = 8, dpi=300)
```    
![](../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov23.png)


```{r eval=FALSE, message=FALSE, warning=FALSE}

pws1<-read.csv("../Output/COV/COVscan_3pop/3pops_top1percent_outlier_regions.cov13_new.csv")
pws1<-pws1[pws1$pop=="PWS" & pws1$window=="100k",]
pws2<-read.csv("../Output/COV/3pops_top1percent_outlier_regions.cov13.csv")
pws2<-pws2[pws2$pop=="PWS"&pws2$window=="100k",]

pws1$vcf<-"PH"
pws2$vcf<-"3Pops"
pws<-rbind(pws1[,c("chrom","start","end","cov13","vcf")],pws2[,c("chrom","start","end","cov13","vcf")])
pws$chrom<-factor(pws$chrom, levels=paste0("chr", 1:26))
ggplot(data=pws,aes(x=start/1000000, y=cov13, color=vcf, fill=vcf))+
    facet_wrap(~chrom)+
    geom_point(position=position_dodge(width = 0.7), alpha=0.5)+xlab("Position (Mb)")+
    ggtitle("PWS COV13")
ggsave("../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov13.png", width = 10, height = 8, dpi=300)
```    

![](../Output/COV/COVscan_3pop/PWS_PH.vs.3Pops_outlier_overlap_cov13.png)






# Interpopulation comparison per time period 
```{r eval=FALSE, message=FALSE, warning=FALSE}
### Interpopulation comparisons
#decode the samples to create the right matrix
cv<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_3pops.csv", header = F)
labs<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_labels_3pops.csv" )
labs<-labs[,-1]
cvm<-data.frame(label=as.vector(t(labs)), cov=as.vector(t(cv)))

#rearrange based on comparions: covariance between populations within the same period
#PopYr Symbols
# PH 1 'PWS', 1991
# PH 2 'PWS', 1996
# PH 3 'PWS', 2006
# PH 4 'PWS', 2017
# PH 5 'SS',  1991
# PH 6 'SS',  1996
# PH 7 'SS',  2006
# PH 8 'SS',  2017
# PH 9 'TB',  1991
# PH 10'TB',  1996
# PH 11'TB',  2006
# PH 12'TB',  2017

Covs<-data.frame(pops=rep(c("PWS.vs.SS", "PWS.vs.TB",  "SS.vs.TB"), times=6),
                 period=c(rep("1991-1996", times=3),rep("1996-2006", times=3), rep("2006-2017", times=3)))

Covs$cov<-c(NA, cvm$cov[cvm$label=="cov(PH: 2-1, PH: 10-9)"],NA,
            cvm$cov[cvm$label=="cov(PH: 3-2, PH: 7-6)"],cvm$cov[cvm$label=="cov(PH: 3-2, PH: 11-10)"], 
            cvm$cov[cvm$label=="cov(PH: 7-6, PH: 11-10)"],
            cvm$cov[cvm$label=="cov(PH: 4-3, PH: 8-7)"],cvm$cov[cvm$label=="cov(PH: 4-3, PH: 12-11)"],cvm$cov[cvm$label=="cov(PH: 8-7, PH: 12-11)"])


#C.I.
cis<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_COV_Interpop_comparison_CIs.csv")
cis<-cis[,-1]
cim<-data.frame(label=as.vector(t(labs)), ci_l=as.vector(t(cis[1:11,])))
cim$ci_h<-as.vector(t(cis[12:22,]))

Covs$ci_l<-as.numeric(c(NA,cim$ci_l[cim$label=="cov(PH: 2-1, PH: 10-9)"],NA,
                      cim$ci_l[cim$label=="cov(PH: 3-2, PH: 7-6)"],cim$ci_l[cim$label=="cov(PH: 3-2, PH: 11-10)"], cim$ci_l[cim$label=="cov(PH: 7-6, PH: 11-10)"],
                      cim$ci_l[cim$label=="cov(PH: 4-3, PH: 8-7)"],cim$ci_l[cim$label=="cov(PH: 4-3, PH: 12-11)"], cim$ci_l[cim$label=="cov(PH: 8-7, PH: 12-11)"]))

Covs$ci_h<-as.numeric(c(NA, cim$ci_h[cim$label=="cov(PH: 2-1, PH: 10-9)"],NA,
                      cim$ci_h[cim$label=="cov(PH: 3-2, PH: 7-6)"],cim$ci_h[cim$label=="cov(PH: 3-2, PH: 11-10)"], cim$ci_h[cim$label=="cov(PH: 7-6, PH: 11-10)"],
                      cim$ci_h[cim$label=="cov(PH: 4-3, PH: 8-7)"],cim$ci_h[cim$label=="cov(PH: 4-3, PH: 12-11)"], cim$ci_h[cim$label=="cov(PH: 8-7, PH: 12-11)"]))

#Barplot
ggplot(Covs, aes(x=period, y=cov, fill=pops))+
    geom_bar(stat="identity",position=position_dodge(width = 0.7), width=0.8)+
    ylab("Covariance")+xlab('')+theme_classic()+
    geom_hline(yintercept = 0,color="gray70", size=0.3)+
    scale_fill_manual(values=cols[c(2,1,3)])+
    theme(legend.title = element_blank())+
    scale_y_continuous(labels = comma)+
    ylim(-0.0013, 0.002)+
    geom_errorbar(aes(ymin=ci_l, ymax=ci_h), width=.2, size=.2, position=position_dodge(width = 0.7))
ggsave("../Output/COV/Interpop_cov_comparison_3Pops_new.png",width = 4.7, height = 3, dpi=300)

#Point plot
ggplot(Covs, aes(x=period, y=cov, color=pops))+
    geom_point(position=position_dodge(width = 0.7), size=4)+
    ylab("Covariance")+xlab('')+theme_classic()+
    geom_hline(yintercept = 0,color="gray70", size=0.3)+
    scale_color_manual(values=cols[c(2,1,3)])+
    theme(legend.title = element_blank())+
    scale_y_continuous(labels = comma)+
    geom_errorbar(aes(ymin=ci_l, ymax=ci_h), width=.2, size=.2, position=position_dodge(width = 0.7))+
    ylim(-0.0013, 0.002)+
     geom_vline(xintercept = c(1.5,2.5), color="gray", size=0.3)
ggsave("../Output/COV/Interpop_cov_comparison_3Pops_new_PointPlot.png",width = 4.7, height = 3, dpi=300)

#line plot
Covs$time<-1
Covs$time[Covs$period=="1996-2006"]<-2
Covs$time[Covs$period=="2006-2017"]<-3
Covs<-Covs[order(Covs$time),]
ggplot(Covs, aes(x=time, y=cov, color=pops, group=pops))+
    geom_point(position=position_dodge(width = 0.7), size=4)+
    geom_path(position=position_dodge(width = 0.7))+
    ylab("Covariance")+xlab('')+theme_classic()+
    geom_hline(yintercept = 0,color="gray70", size=0.3)+
    scale_color_manual(values=cols[c(2,1,3)])+
    theme(legend.title = element_blank())+
    scale_y_continuous(labels = comma)+
    geom_errorbar(aes(ymin=ci_l, ymax=ci_h), width=.2, size=.2, position=position_dodge(width = 0.7))+
    ylim(-0.0013, 0.002)+
     geom_vline(xintercept = c(1.5,2.5), color="gray", size=0.3)+
    scale_x_continuous(breaks=c(1,2,3), labels = c("1991-1996","1996-2006","2006-2017"))
ggsave("../Output/COV/Interpop_cov_comparison_3Pops_new_LinePlot.png",width = 4.7, height = 3, dpi=300)

```

![](../Output/COV/Interpop_cov_comparison_3Pops_new_PointPlot.png)

## Longer time period
```{r eval=FALSE, message=FALSE, warning=FALSE}
## Longer time-period
Covs2<-data.frame(pops=rep(c("PWS.vs.SS", "PWS.vs.TB",  "SS.vs.TB"), times=3),
                 period=c(rep("1991-2006", times=3),rep("1991-2017", times=3),rep("1996-2017", times=3)))

cv1<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_1991-2006.csv", header = F)
labs1<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_labels_1991-2006.csv" )
labs1<-labs1[,-1]
cvm1<-data.frame(label=as.vector(t(labs1)), cov=as.vector(t(cv1)))

cv2<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_1991-2017.csv", header = F)
labs2<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_labels_1991-2017.csv" )
labs2<-labs2[,-1]
cvm2<-data.frame(label=as.vector(t(labs2)), cov=as.vector(t(cv2)))

cv3<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_1996-2017.csv", header = F)
labs3<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_Covs_interPopulations_100k_labels_1996-2017.csv" )
labs3<-labs3[,-1]
cvm3<-data.frame(label=as.vector(t(labs3)), cov=as.vector(t(cv3)))

Covs2$cov<-c(NA, cvm1$cov[cvm1$label=="cov(PH: 2-1, PH: 4-3)"], NA,
             NA, cvm2$cov[cvm2$label=="cov(PH: 2-1, PH: 4-3)"], NA,
             cvm3$cov[cvm3$label=="cov(PH: 2-1, PH: 4-3)"], cvm3$cov[cvm3$label=="cov(PH: 2-1, PH: 6-5)"], cvm3$cov[cvm3$label=="cov(PH: 4-3, PH: 6-5)"])

#C.I.
cis1<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_COV_Interpop_comparison_CIs_1991-2006.csv")
cis1<-cis1[,-1]
cis2<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_COV_Interpop_comparison_CIs_1991-2017.csv")
cis2<-cis2[,-1]
cis3<-read.csv("~/Projects/Pacherring_Vincent/MD7000/GW_COV_Interpop_comparison_CIs_1996-2017.csv")
cis3<-cis3[,-1]

#cim<-data.frame(label=as.vector(t(labs)), ci_l=as.vector(t(cis1[1:4,])))
#cim$ci_h<-as.vector(t(cis[12:22,]))

Covs2$ci_l<-as.numeric(c(NA,cis1[1,3],NA,
                        NA,cis2[1,3],NA,
                      cis3[1,3],cis3[1,5],cis3[3,5]))

Covs2$ci_h<-as.numeric(c(NA,cis1[4,3],NA,
                        NA,cis2[4,3],NA,
                      cis3[6,3],cis3[6,5],cis3[8,5]))


ggplot(Covs2, aes(x=period, y=cov, fill=pops))+
    geom_bar(stat="identity",position=position_dodge(width = 0.7), width=0.8)+
    ylab("Covariance")+xlab('')+theme_classic()+
    geom_hline(yintercept = 0,color="gray70", size=0.3)+
    scale_fill_manual(values=cols[c(2,1,3)])+
    theme(legend.title = element_blank())+scale_y_continuous(labels = comma)+
    ylim(-0.0013, 0.002)+
    geom_errorbar(aes(ymin=ci_l, ymax=ci_h), width=.2, size=.2, position=position_dodge(width = 0.7))
ggsave("../Output/COV/Interpop_cov_comparison_3Pops_LonogerPeriod.png",width = 4.7, height = 3, dpi=300)

ggplot(Covs2, aes(x=period, y=cov, color=pops))+
    geom_point(position=position_dodge(width = 0.7), size=4)+
    ylab("Covariance")+xlab('')+theme_classic()+
    geom_hline(yintercept = 0,color="gray70", size=0.3)+
    scale_color_manual(values=cols[c(2,1,3)])+
    theme(legend.title = element_blank())+
    scale_y_continuous(labels = comma)+
     ylim(-0.0013, 0.002)+
    geom_errorbar(aes(ymin=ci_l, ymax=ci_h), width=.2, size=.2, position=position_dodge(width = 0.7))+
    geom_vline(xintercept = c(1.5,2.5), color="gray", size=0.3)
ggsave("../Output/COV/Interpop_cov_comparison_3PopsLonogerPeriod_PointPlot.png",width = 4.7, height = 3, dpi=300)


```
![](../Output/COV/Interpop_cov_comparison_3PopsLonogerPeriod_PointPlot.png)

# Focused freq analysis
## rel gene (chr13: 23070000 - 23080000)

```{r eval=FALSE, message=FALSE, warning=FALSE}
pops<-c("PWS91","PWS96","PWS07","PWS17")
yr<-c(1991,1996,2007,2017)
maf<-data.frame()
for (i in 1:4){
    af<-read.table(paste0("../Data/new_vcf/AF/",pops[i],".mafs"),sep="\t", header = T)
    af<-af[af$chromo=="chr13"&af$position>=23070000&af$position<=23080000,]
    af$year<-yr[i]
    maf<-rbind(maf,af)
}
#write.csv(maf,"../Output/COV/COVscan_3pop/AF_maf_chr13_23Mb.csv")

ggplot(maf, aes(x=year, y=knownEM, color=factor(position)))+
    geom_point()+
    geom_path()+ggtitle("MAF (ANGSD)")+
    ylab("maf")+
    theme(legend.title=element_blank())
ggsave("../Output/COV/COVscan/AF_ch13_23.07-23.08_angsd.png", width = 5, height=3, dpi=300)


AF<-data.frame()
for (i in 1:4){
    af<-read.table(paste0("../Data/new_vcf/AF/",pops[i],"_maf05_af.frq"),header = FALSE, skip=1, col.names = c("chr","pos","n_allele","n_sample","MajorAF","MAF"))
    af<-af[af$chr=="chr13"&af$pos>=23070000&af$pos<=23080000,]
    af$year<-yr[i]
    af$maf<-substr(af$MAF, 3,10)
    af$maf<-as.numeric(af$maf)
    af<-af[,c("chr","pos","maf","year")]
    AF<-rbind(AF,af)
}
ggplot(AF, aes(x=year, y=maf, color=factor(pos)))+
    geom_point()+
    geom_path()+ggtitle("MAF (vcftools)")+
    theme(legend.title=element_blank())
ggsave("../Output/COV/COVscan/AF_ch13_23.07-23.08.png", width = 5, height=3, dpi=300)


###TB
pops<-c("TB91","TB96","TB06","TB17","PWS91","PWS96","PWS07","PWS17","SS96","SS06","SS17")
yr<-c(1991,1996,2006,2017,1991,1996,2007,2017,1996,2006,2017)
maf<-data.frame()
for (i in 1:length(pops)){
    af<-read.table(paste0("../Data/new_vcf/AF/",pops[i],".mafs"),sep="\t", header = T)
    af<-af[af$chromo=="chr13"&af$position>=23070000&af$position<=23080000,]
    af$year<-yr[i]
    af$pop<-sub("\\d\\d","", pops[i])
    maf<-rbind(maf,af)
}
#write.csv(maf,"../Output/COV/COVscan_3pop/AF_maf_chr13_23Mb.csv")
ggplot(maf, aes(x=year, y=knownEM, color=pop))+
    facet_wrap(~factor(position))+
    geom_point()+
    geom_path()+ggtitle("Chr13 (rel gene)")+
    ylab("maf")+
    theme(legend.title=element_blank())
ggsave("../Output/COV/COVscan/AF_ch13_23.07-23.08_angsd.png", width = 5, height=3, dpi=300)




```
![](../Output/COV/COVscan/AF_ch13_23.07-23.08_angsd.png)





```{r eval=FALSE, message=FALSE, warning=FALSE}




```

```{r eval=FALSE, message=FALSE, warning=FALSE}
```

